src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageParser.g4

```diff
@@ -864,9 +864,19 @@ arguments
     : LPAREN expressionList? RPAREN
     ;
 
+/**
+ * We do for patterns as we do for expressions; namely we have one parent
+ * 'PATTERN_DEF' node, then have all nested pattern definitions inside of
+ * the parent node.
+ */
 pattern
-    : guardedPattern
-    | primaryPattern
+    : innerPattern
+    ;
+
+innerPattern
+    : g=guardedPattern
+    | r=recordPattern
+    | p=primaryPattern
```

Comment:
what for "g=" like assignments ?

Reply:
This is just me being lazy while developing the grammar/visitor, this just lets you reference the rule by `ctx.g` instead of `ctx.guardedPattern`. This will be removed in final iteration.